test/ResidualTest.cc

```diff
@@ -0,0 +1,460 @@
+        input_aligned_ = (uint8_t *)eb_aom_memalign(8, test_size_);
+        pred_aligned_ = (uint8_t *)eb_aom_memalign(8, test_size_);
+        residual1_aligned_ = (int16_t *)eb_aom_memalign(16, 2 * test_size_);
+        residual2_aligned_ = (int16_t *)eb_aom_memalign(16, 2 * test_size_);
+        input16bit_aligned_ = (uint16_t *)eb_aom_memalign(16, 2 * test_size_);
+        pred16bit_aligned_ = (uint16_t *)eb_aom_memalign(16, 2 * test_size_);
+        residual_aligned_ = nullptr;
+        is_16bit_Use_ = true;
+        memset(
+            residual1_aligned_, 0, test_size_ * sizeof(residual1_aligned_[0]));
+        memset(
+            residual2_aligned_, 0, test_size_ * sizeof(residual2_aligned_[0]));
+    }
+
+    ResidualTestBase(uint32_t area_width, uint32_t area_height,
+                     TestPattern test_pattern,
+                     SUB_SAMPLED_TEST_FUNC sub_sampled_test_func) {
+        area_width_ = area_width;
+        area_height_ = area_height;
+        test_pattern_ = test_pattern;
+        sub_sampled_test_func_ = sub_sampled_test_func;
+        input_stride_ = pred_stride_ = residual_stride_ = MAX_PU_SIZE;
+        test_size_ = MAX_PU_SIZE * MAX_PU_SIZE;
+        input_aligned_ = (uint8_t *)eb_aom_memalign(8, test_size_);
+        pred_aligned_ = (uint8_t *)eb_aom_memalign(8, test_size_);
+        residual1_aligned_ = (int16_t *)eb_aom_memalign(16, 2 * test_size_);
+        residual2_aligned_ = (int16_t *)eb_aom_memalign(16, 2 * test_size_);
+        residual_aligned_ = nullptr;
+        memset(
+            residual1_aligned_, 0, test_size_ * sizeof(residual1_aligned_[0]));
+        memset(
+            residual2_aligned_, 0, test_size_ * sizeof(residual2_aligned_[0]));
+    }
+
+    void TearDown() override {
+        if (input_aligned_)
+            eb_aom_free(input_aligned_);
+        if (pred_aligned_)
+            eb_aom_free(pred_aligned_);
+        if (!is_16bit_Use_)
+            eb_aom_free(input16bit_aligned_);
+        if (!is_16bit_Use_)
+            eb_aom_free(pred16bit_aligned_);
+        if (residual1_aligned_)
+            eb_aom_free(residual1_aligned_);
+        if (residual2_aligned_)
+            eb_aom_free(residual2_aligned_);
+        if (residual_aligned_)
+            eb_aom_free(residual_aligned_);
+    }
+
+  protected:
+    void prepare_data() {
+        const int16_t mask = (1 << 8) - 1;
+        SVTRandom rnd(0, mask);
+        switch (test_pattern_) {
+        case REF_MIN: {
+            memset(input_aligned_, 0, test_size_ * sizeof(input_aligned_[0]));
+            memset(pred_aligned_, 0, test_size_ * sizeof(pred_aligned_[0]));
+            break;
+        }
+        case REF_MAX: {
+            memset(
+                input_aligned_, mask, test_size_ * sizeof(input_aligned_[0]));
+            memset(pred_aligned_, 0, test_size_ * sizeof(pred_aligned_[0]));
+            break;
+        }
+        case REF_RANDOM: {
+            for (int i = 0; i < test_size_; i++) {
```

Comment:
GNU warnings: warning: comparison of integers of different signs: 'int' and 'uint32_t' (aka 'unsigned int') [-Wsign-compare]